### ISSUE TITLE
Unable to create endpoint with `-t disabled` in kernel 4.9.212

### ISSUE THREAD

**Issue by aanm** [cilium/cilium#11107]
I've bisected when the regression was introduced and it seems it is because of
https://github.com/cilium/cilium/commit/307c3e5b4106bf385a6b19db6f246d43f3ada40f

Cilium arguments 
`--debug --pprof --enable-k8s-event-handover --k8s-require-ipv4-pod-cidr --auto-direct-node-routes --k8s-kubeconfig-path /var/lib/cilium/cilium.kubeconfig --kvstore etcd --kvstore-opt etcd.config=/var/lib/cilium/etcd-config.yml -t disabled`

Deploy the following deployment:
```
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: test
  name: test
spec:
  progressDeadlineSeconds: 600
  replicas: 6
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      app: test
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - imagePullPolicy: IfNotPresent
        name: test
        image: docker.io/library/busybox:1.29
        command: [ 'sleep', '1000h' ]
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
```

`cilium endpoint list`
```
3545       Disabled           Disabled          35295      k8s:app=test                                      f00d::a11:0:0:f680   10.17.47.120    regenerating            
                                                           k8s:io.cilium.k8s.policy.cluster=default                                                                       
                                                           k8s:io.cilium.k8s.policy.serviceaccount=default                                                                
                                                           k8s:io.kubernetes.pod.namespace=default 
```

```
$ cilium endpoint log 3545
Timestamp              Status    State                   Message
2020-04-22T14:06:40Z   OK        waiting-to-regenerate   retrying regeneration
2020-04-22T14:06:40Z   Failure   ready                   Error regenerating endpoint: Failed to load tc filter: exit status 1
2020-04-22T14:06:40Z   OK        ready                   Completed endpoint regeneration with no pending regeneration requests
2020-04-22T14:06:19Z   OK        regenerating            Regenerating endpoint: updated security labels
2020-04-22T14:05:42Z   OK        waiting-to-regenerate   Triggering regeneration due to new identity
2020-04-22T14:05:42Z   OK        ready                   Set identity for this endpoint
2020-04-22T14:05:42Z   OK        waiting-for-identity    Endpoint creation
```

There are no relevant logs in cilium

I still need to retry it in more recent kernels.

**Comment by aanm** [comment]
Further debugging it seems this is cause by complexity limit.
```
$ sudo tc filter add dev foo0 ingress bpf da obj /var/run/cilium/state/templates/1ae4f2dcd7cd3fa70bf105085c431914adf5d34f/bpf_lxc.o sec from-container

Prog section '2/10' rejected: Argument list too long (7)!
 - Type:         3
 - Attach Type:  0
 - Instructions: 3409 (0 over limit)
 - License:      GPL

Verifier analysis:

Skipped 4230922 bytes, use 'verb' option for the full verbose log.
[...]
 R8=inv,min_value=0,max_value=0 R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3151: (b7) r6 = 0
3152: (63) *(u32 *)(r10 -64) = r6
3153: (bf) r3 = r10
3154: (07) r3 += -64
3155: (b7) r8 = 1
3156: (b7) r1 = 0
3157: (b7) r2 = 0
3158: (b7) r4 = 1
3159: (b7) r5 = 0
3160: (85) call 28
3161: safe

from 3150 to 3206: R1=inv,min_value=0,max_value=0 R2=imm0,min_value=0,max_value=0 R6=ctx R7=inv R8=inv,min_value=0,max_value=0 R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3206: (bf) r0 = r7
3207: (95) exit

from 3146 to 3148: R1=inv32,min_value=2,max_value=2 R2=imm1,min_value=1,max_value=1 R6=ctx R7=inv R8=inv,min_value=0,max_value=0 R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3148: (77) r1 >>= 31
3149: (4f) r1 |= r2
3150: (15) if r1 == 0x0 goto pc+55
 R1=inv R2=imm1,min_value=1,max_value=1 R6=ctx R7=inv R8=inv,min_value=0,max_value=0 R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3151: (b7) r6 = 0
3152: (63) *(u32 *)(r10 -64) = r6
3153: (bf) r3 = r10
3154: (07) r3 += -64
3155: (b7) r8 = 1
3156: (b7) r1 = 0
3157: (b7) r2 = 0
3158: (b7) r4 = 1
3159: (b7) r5 = 0
3160: (85) call 28
3161: safe

from 3150 to 3206: R1=inv,min_value=0,max_value=0 R2=imm1,min_value=1,max_value=1 R6=ctx R7=inv R8=inv,min_value=0,max_value=0 R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3206: (bf) r0 = r7
3207: (95) exit

from 2911 to 3142: safe

from 2478 to 2508: R0=inv R2=imm0,min_value=0,max_value=0 R6=inv R7=imm0,min_value=0,max_value=0 R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
2508: (18) r7 = 0xffffff65
2510: (05) goto pc+631
3142: (bf) r1 = r7
3143: (67) r1 <<= 32
3144: (77) r1 >>= 32
3145: (b7) r2 = 1
3146: (15) if r1 == 0x2 goto pc+1
 R0=inv R1=inv32 R2=imm1,min_value=1,max_value=1 R6=inv R7=inv R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3147: (b7) r2 = 0
3148: (77) r1 >>= 31
3149: (4f) r1 |= r2
3150: (15) if r1 == 0x0 goto pc+55
 R0=inv R1=inv R2=imm0,min_value=0,max_value=0 R6=inv R7=inv R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3151: (b7) r6 = 0
3152: (63) *(u32 *)(r10 -64) = r6
3153: (bf) r3 = r10
3154: (07) r3 += -64
3155: (b7) r8 = 1
3156: (b7) r1 = 0
3157: (b7) r2 = 0
3158: (b7) r4 = 1
3159: (b7) r5 = 0
3160: (85) call 28
3161: (79) r2 = *(u64 *)(r10 -184)
3162: (63) *(u32 *)(r2 +52) = r9
3163: (18) r1 = 0x2
3165: (63) *(u32 *)(r2 +48) = r1
3166: (b7) r1 = 2
3167: (63) *(u32 *)(r2 +64) = r1
3168: (63) *(u32 *)(r2 +56) = r7
3169: (63) *(u32 *)(r2 +60) = r6
3170: (61) r9 = *(u32 *)(r2 +0)
3171: (7b) *(u64 *)(r10 -56) = r6
3172: (7b) *(u64 *)(r10 -64) = r6
3173: (b7) r1 = 512
3174: (7b) *(u64 *)(r10 -104) = r1
3175: (87) r7 neg 0
3176: (73) *(u8 *)(r10 -104) = r7
3177: (bf) r2 = r10
3178: (07) r2 += -104
3179: (18) r1 = 0xffff8ce35a940d80
3181: (85) call 1
3182: safe

from 3150 to 3206: R0=inv R1=inv,min_value=0,max_value=0 R2=imm0,min_value=0,max_value=0 R6=inv R7=inv R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3206: (bf) r0 = r7
3207: (95) exit

from 3146 to 3148: R0=inv R1=inv32,min_value=2,max_value=2 R2=imm1,min_value=1,max_value=1 R6=inv R7=inv R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3148: (77) r1 >>= 31
3149: (4f) r1 |= r2
3150: (15) if r1 == 0x0 goto pc+55
 R0=inv R1=inv R2=imm1,min_value=1,max_value=1 R6=inv R7=inv R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3151: (b7) r6 = 0
3152: (63) *(u32 *)(r10 -64) = r6
3153: (bf) r3 = r10
3154: (07) r3 += -64
3155: (b7) r8 = 1
3156: (b7) r1 = 0
3157: (b7) r2 = 0
3158: (b7) r4 = 1
3159: (b7) r5 = 0
3160: (85) call 28
3161: safe

from 3150 to 3206: R0=inv R1=inv,min_value=0,max_value=0 R2=imm1,min_value=1,max_value=1 R6=inv R7=inv R8=ctx R9=inv R10=fp fp-192=map_ptr fp-184=ctx
3206: (bf) r0 = r7
3207: (95) exit

from 2434 to 2437: safe

from 2427 to 2435: safe

from 2413 to 2415: R0=inv,min_value=0 R1=inv48 R2=inv56,min_value=0,max_value=0 R3=imm0,min_value=0,max_value=0 R4=inv,min_value=0,max_value=0 R5=inv,max_value=0 R6=inv,min_value=0,max_value=0 R7=imm0,min_value=0,max_value=0 R8=imm60,min_value=60,max_value=60 R9=inv R10=fp fp-192=map_ptr fp-184=ctx
2415: (4f) r1 |= r2
2416: (6b) *(u16 *)(r10 -28) = r1
2417: (85) call 5
2418: (37) r0 /= 1000000000
BPF program is too large. Proccessed 98305 insn

Error filling program arrays!
Unable to load program

```

**Comment by pchaigno** [comment]
Do you know how far from the complexity limit we are? Alternatively, could you attach the object file to this thread?

**Comment by borkmann** [comment]
> Do you know how far from the complexity limit we are? Alternatively, could you attach the object file to this thread?

Yes, from complexity limit. Looks like it didn't get caught by our CI on 4.9 kernels.

**Comment by aanm** [comment]
@pchaigno 

[bpf_lxc.zip](https://github.com/cilium/cilium/files/4516908/bpf_lxc.zip)

**Comment by pchaigno** [comment]
We're 35,351 instructions over the limit:
```
Prog section '2/10' loaded (29)!
processed 133656 insns
```
[The full dump on that modified 4.9.212 kernel](https://github.com/cilium/cilium/files/4517804/bpf_lxc-complex-issue.all.txt) is attached. In the `#sig-datapath` meeting, we've discussed a couple mitigations which could be applied here. In addition, I annotated the program bytecode dump with the number of times the verifier analyzed each instructions and it looks like we could reduce complexity by improving/adding pruning points with `relax_verifier()`.
